packages/astro/src/core/util.ts

```diff
@@ -74,12 +74,21 @@ export function resolveDependency(dep: string, astroConfig: AstroConfig) {
 }
 
 /**
- * Vite-ify URL
- * Given a file URL, return an ID that matches Vite’s module graph. Needed for resolution and stack trace fixing.
- * Must match the following format:
+ * Convert file URL to Vite URL
+ * ! Warning: this works in dev server, but may not always be present in Vite’s module graph
  * Linux/Mac:  /@fs/Users/astro/code/my-project/src/pages/index.astro
  * Windows:    /@fs/C:/Users/astro/code/my-project/src/pages/index.astro
  */
-export function viteifyURL(filePath: URL): string {
+export function viteURL(filePath: URL): string {
   return `/@fs${slash(fileURLToPath(filePath)).replace(/^\/?/, '/')}`;
 }
+
+/**
+ * Convert file URL to Vite ID
+ * ! Warning: this is almost always present in Vite‘s module graph, but doesn’t work in dev server
+ * Linux/Mac:  /Users/astro/code/my-project/src/pages/index.astro
+ * Windows:    C:/Users/astro/code/my-project/src/pages/index.astro
+ */
+export function viteID(filePath: URL): string {
```

Comment:
1/2 The core of this fix is pretty simple: rather than one “function to rule them all” we split it up into `viteURL` that helps with `moduleGraph.urlToModuleMap` and `viteID` that helps with `moduleGraph.idToModuleMap` …